actors/evm/src/interpreter/precompiles/parameter.rs

```diff
@@ -18,212 +19,197 @@ impl From<GroupError> for PrecompileError {
+    pub fn drop_zeros<const S: usize>(&mut self) -> Result<(), PrecompileError> {
+        let split = S.min(self.slice.len());
+        let (a, b) = self.slice.split_at(split);
+        self.slice = b;
+        if a.iter().all(|&i| i == 0) {
+            Ok(())
+        } else {
+            Err(PrecompileError::InvalidInput)
+        }
     }
 
-    pub fn next_padded(&mut self) -> [T; CHUNK_SIZE]
-    where
-        T: Default,
-    {
-        if self.chunks.len() > 0 {
-            self.next().copied().unwrap_or([T::default(); CHUNK_SIZE])
-        } else if self.exhausted() {
-            [T::default(); CHUNK_SIZE]
+    /// Read a single byte, or 0 if there's no remaining input.
+    pub fn read_byte(&mut self) -> u8 {
+        if let Some((&first, rest)) = self.slice.split_first() {
+            self.slice = rest;
+            first
         } else {
-            self.exhausted = true;
-            let mut buf = [T::default(); CHUNK_SIZE];
-            let remainder = self.chunks.remainder();
-            buf[..remainder.len()].copy_from_slice(remainder);
-            buf
+            0
         }
     }
 
-    pub fn exhausted(&self) -> bool {
-        self.exhausted
+    /// Read a fixed number of bytes from the input, zero-padding as necessary.
+    pub fn read_fixed<const S: usize>(&mut self) -> [u8; S] {
+        let mut out = [0u8; S];
+        let split = S.min(self.slice.len());
+        let (a, b) = self.slice.split_at(split);
+        self.slice = b;
+        out[..split].copy_from_slice(a);
+        out
     }
 
-    pub fn remaining_len(&self) -> usize {
-        if self.exhausted {
-            0
+    /// Read input and pad up to `len`.
+    pub fn read_padded(&mut self, len: usize) -> Cow<'a, [u8]> {
+        if len <= self.slice.len() {
+            let (a, b) = self.slice.split_at(len);
+            self.slice = b;
+            Cow::Borrowed(a)
         } else {
-            self.chunks.len() * CHUNK_SIZE + self.chunks.remainder().len()
+            let mut buf = Vec::with_capacity(len);
+            buf.extend_from_slice(self.slice);
+            buf.resize(len, 0);
+            self.slice = &[];
+            Cow::Owned(buf)
         }
     }
 
-    pub fn chunks_read(&self) -> usize {
-        let total_chunks = self.slice.len() / CHUNK_SIZE;
-        let unread_chunks = self.chunks.len();
-        total_chunks - unread_chunks
-    }
-
-    // remaining unpadded slice of unread items
-    pub fn remaining_slice(&self) -> &[T] {
-        let start = self.slice.len() - self.remaining_len();
-        &self.slice[start..]
+    /// Read a bigint from the input, and pad up to `len`.
+    pub fn read_biguint(&mut self, len: usize) -> BigUint {
+        let split = len.min(self.slice.len());
+        let (a, b) = self.slice.split_at(split);
+        self.slice = b;
+        let mut int = BigUint::from_bytes_be(a);
+        if split < len {
+            int <<= (len - split) * 8;
+        }
+        int
```

Comment:
could you add some comments to better describe what this is doing? a bit hard to follow.

Reply:
Done.